libp2p/libp2pendpoint_test.go

```diff
@@ -425,7 +430,7 @@ func TestReqTimeout(t *testing.T) {
 	go func() {
 		// timeout is queued in the scheduler 0
 		for !scheds[0].RunOne(ctx) {
-			time.Sleep(1 * time.Millisecond)
+			time.Sleep(10 * time.Millisecond)
```

Comment:
This is problematic and means we don't have deterministic tests. If we're tweaking test timings then we're not in a better position than using raw goroutines. 

We should try and remove sleeps from all tests. As a side issue, why aren't these sleeps using the scheduler's clock?

Reply:
@iand unfortunately we cannot get deterministic testing when using go-libp2p. When using go-libp2p to send messages over the network stack, multiple go routines are at play, and we cannot use the scheduler to obtain sequentiality.

> We should try and remove sleeps from all tests

I wish we were able to do it, but we need to _wait_ for the message to be passed from one go routine to another over the network stack.

This test is testing the request timeout. `A` is sending a request to `B`. No action is run in `B`'s scheduler until `A`'s request times out. After `A` has timed out, `B` will send a late reply, and `A` should discard it. However, sometimes `B`'s scheduler has no action to run after `A`'s timeout of 100 ms (!!!).

There are 3 go routines at play:
1. The single worker, running both `A` and `B` schedulers.
2. `A`'s go-libp2p go routine sending the request.
3. `B`'s go-libp2p go routine listening at incoming requests.

It means that either (2) is not sending the request within 100ms or that (3) isn't processing the request within 100ms (, or that the request got lost in the local network stack, but it seems unlikely). It is very hard to understand what is wrong, because I cannot reproduce the failure on my machine (that is certainly faster than GH CI), and it only happens once in a while on GH CI.

I will try to add verbose when the test fails, and run it many times in GH CI.

> As a side issue, why aren't these sleeps using the scheduler's clock?

You are right! In this case however it won't solve the flakiness, because the scheduler's clock is a _real clock_.